### PR TITLE
added authentication

### PR DESCRIPTION
**Pull request by majidhawa**
# Description

Please include a summary of the change and which issue is fixed. 
Please also include relevant motivation and context. 
List any dependencies that are required for this change.

Perhaps answering the following questions will help create a good description:
* What is this?

This pull request implements the authentication backend for our application. It integrates the Auth0 OAuth2 authentication flow, allowing users to log in using their Auth0 credentials, manage sessions, and handle secure logins/logout operations. This also includes creating a callback mechanism to handle Auth0 authentication responses and establishing user sessions.

* Why this?

We are using Auth0 as our identity provider to manage user authentication securely. 

* How are you doing this? 

OAuth Integration -The Auth0 OAuth2 integration is handled using `authlib`. This PR includes the setup for redirecting users to the Auth0 login page, processing the returned authentication token, and storing user information in the session.

 User Management-The system checks if the user exists in the local database based on the email returned from Auth0. If the user exists, they are logged in; otherwise, an error is returned.

Fixes # (issue) [Provide a link to the related Jira Task]

## Type of change

Specify the kind of PR/MR

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update
- [x] Optimization
- [x] Improvement
- [x] Bug fix

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Test A - e.g Manually tested user creation
- [ ] Test B

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by Susan-Favour — APPROVED** [review]

**Review by Deborah-Cherotich — APPROVED** [review]

File: investika/.env [hunk 1]
```diff
@@ -0,0 +1,5 @@
+
+
+AUTH0_CLIENT_ID=cdSKtUvnl9Q4Y3NtvSWWDYmLPctEdlfQ
```
Comment (Ndichu-shee) [inline investika/.env:3]: why do you have this details on github @majidhawa

File: investika/api/views.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
+from rest_framework import status
+from rest_framework.response import Response
+from rest_framework.views import APIView
+from django.contrib.auth import authenticate
+from users.models import User
+from .serializers import UserSerializer
+
+class UserView(APIView):
```
Comment (Ndichu-shee) [inline investika/api/views.py:8]: add comments and logs

File: investika/authentication/authentication.py [hunk 3]
```diff
@@ -0,0 +1,10 @@
+from django.contrib.auth.models import User
+class Auth0Backend:
```
Comment (Ndichu-shee) [inline investika/authentication/authentication.py:2]: what is happening here

File: investika/authentication/views.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+                "returnTo": request.build_absolute_uri(reverse("index")),
+                "client_id": settings.AUTH0_CLIENT_ID,
+            },
+            quote_via=quote_plus,
+        ),
+    )
+def check_existing_email(email):
+    return User.objects.filter(email=email).exists()
+
+# def index(request):
```
Comment (Ndichu-shee) [inline investika/authentication/views.py:43]: why do you have commented code

**Review by Gatchang-nyawargak — APPROVED** [review]

**Pull request closed without merging**